pallets/passkey/src/lib.rs

```diff
@@ -131,28 +145,42 @@ pub mod module {
 	#[pallet::validate_unsigned]
 	impl<T: Config> ValidateUnsigned for Pallet<T>
 	where
-		<T as frame_system::Config>::RuntimeCall: Dispatchable<Info = DispatchInfo>,
+		<T as frame_system::Config>::RuntimeCall:
+			From<Call<T>> + Dispatchable<Info = DispatchInfo, PostInfo = PostDispatchInfo>,
 	{
 		type Call = Call<T>;
 		fn validate_unsigned(_source: TransactionSource, call: &Self::Call) -> TransactionValidity {
 			let payload = Self::filter_valid_calls(&call)?;
 			Self::validate_signatures(&payload)?;
-
+			let tx_charge = ChargeTransactionPayment::<T>(
+				payload.passkey_call.account_id.clone(),
+				call.clone(),
+			);
+			tx_charge.validate()?;
 			let nonce_check = PasskeyNonce::new(payload.passkey_call.clone());
 			nonce_check.validate()
 		}
 
 		fn pre_dispatch(call: &Self::Call) -> Result<(), TransactionValidityError> {
+			let payload = Self::filter_valid_calls(&call)?;
+			let tx_charge = ChargeTransactionPayment::<T>(
+				payload.passkey_call.account_id.clone(),
+				call.clone(),
+			);
+			tx_charge.pre_dispatch()?;
+
```

Comment:
Additionally, it seems that we are calling validate_unsigned inside pre_dispatch, which might be redundant.

Reply:
Yeah i was wondering why we need pre_dispatch here, the actual trait `ValidatedUnsigned` provides a default implementation of `pre_dispatch` that calls `validate_unsigned` so it feels redundant!

https://github.com/paritytech/polkadot-sdk/blob/18228a9bdbe5b560a39a31810fdf2e3fba59a40d/substrate/primitives/runtime/src/traits.rs#L1781


Thought?

Reply:
It appears that if we only write validate_unsigned, it will be called again during pre-dispatch, which seems reasonable. However, since we are not using the default implementation and are instead calling pre_dispatch, which in turn calls validate, it seems fine to avoid calling validate_unsigned again.

Additionally, there are multiple ways to approach this, and I wanted to highlight that we are performing multiple state transitions. It’s important to note that calling it twice might unintentionally result in some stateful transactions.